File: 프로그래머스/Level3/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

vector<int> solution(vector<string> operations) {
    vector<int> answer;
    
    priority_queue<int> max_pq;
    for(int i=0; i<operations.size(); i++){
        char cmd = operations[i][0];
        int num = stoi(operations[i].substr(2));
        
        if(cmd == 'I')
            max_pq.push(num);
        else if(cmd == 'D'){
            if(num == 1){
                if(!max_pq.empty())
                    max_pq.pop();
            }
            else{
                if(!max_pq.empty()){
                    priority_queue<int, vector<int>, greater<int> > min_pq;
                    while(!max_pq.empty()){
                        min_pq.push(max_pq.top());
                        max_pq.pop();
                    }
                    min_pq.pop();
                
                    while(!min_pq.empty()){
                        max_pq.push(min_pq.top());
                        min_pq.pop();
                    }
                }
            }
        }
    }
    
    if(max_pq.size() > 0){
        int max_value = max_pq.top();
        while(max_pq.size() > 1){
            max_pq.pop();
        }
        int min_value = max_pq.top();
        
        answer.push_back(max_value);
        answer.push_back(min_value);
    }
    else{
        answer.push_back(0);
        answer.push_back(0);
    }
    return answer;
}

File: 프로그래머스/Level3/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        Map<Integer, Integer> map = new HashMap<>();
        char cmd;
        int num;
        
        for(String op : operations) {
            cmd = op.charAt(0);
            num = Integer.parseInt(op.substring(2));
            
            if(cmd == 'I') {
                minHeap.add(num);
                maxHeap.add(num);
                map.put(num, map.getOrDefault(num, 0) + 1);
            } else {
                if(num > 0) {
                    while(!maxHeap.isEmpty()) {
                        int top = maxHeap.poll();
                        if(map.get(top) != 0) {
                            map.put(top, map.get(top) - 1);
                            break;
                        } 
                    }
                } else {
                    while(!minHeap.isEmpty()) {
                        int top = minHeap.poll();
                        if(map.get(top) != 0) {
                            map.put(top, map.get(top) -  1);
                            break;
                        }
                    }
                }
            }
        }
        
        return getAnswer(minHeap, maxHeap, map);
    }
    
    int[] getAnswer(Queue<Integer> minHeap, Queue<Integer> maxHeap, Map<Integer, Integer> map) {
        int maxVal = 0;
        int minVal = 0;
        int top;
        while(!maxHeap.isEmpty()) {
            top = maxHeap.poll();
            if(map.get(top) != 0) {
                maxVal = top;
                break;
            }
        }
        while(!minHeap.isEmpty()) {
            top = minHeap.poll();
            if(map.get(top) != 0) {
                minVal = top;
                break;
            }
        }
        return new int[] {maxVal, minVal};
    }
}
